testbed/tests/test_font.py

```diff
@@ -0,0 +1,131 @@
+):
+    """If an unknown font is requested, the system font is used as a fallback."""
+    widget_probe.assert_font_family(SYSTEM)
+    widget.style.font_family = "unknown"
+    await widget_probe.redraw("Falling back to system font")
+
+    assert "using system font as a fallback" in capsys.readouterr().out
+
+
+async def test_font_options(widget: toga.Label, widget_probe: Any):
+    """Every combination of weight, style and variant can be used on a font."""
+    for font_weight in widget_probe.FONT_WEIGHTS:
+        for font_style in widget_probe.FONT_STYLES:
+            for font_variant in widget_probe.FONT_VARIANTS:
+                widget.style.font_style = font_style
+                widget.style.font_variant = font_variant
+                widget.style.font_weight = font_weight
+                await widget_probe.redraw(
+                    f"Using a {font_weight} {font_style} {font_variant} font"
+                )
+
+                assert widget_probe.font.weight == font_weight
+                assert widget_probe.font.style == font_style
+                assert widget_probe.font.variant == font_variant
+
+
+@pytest.mark.parametrize(
+    "font_family,font_path,font_kwargs",
+    [
+        # OpenType font, no options
+        (
+            "Font Awesome 5 Free Solid",
+            "resources/fonts/Font Awesome 5 Free-Solid-900.otf",
+            {},
+        ),
+        # TrueType font, no options
+        ("Endor", "resources/fonts/ENDOR___.ttf", {}),
+        # Font with weight property
+        ("Roboto", "resources/fonts/Roboto-Bold.ttf", {"weight": BOLD}),
+        # Font with style property
+        ("Roboto", "resources/fonts/Roboto-Italic.ttf", {"style": ITALIC}),
+        # Font with multiple properties
+        (
+            "Roboto",
+            "resources/fonts/Roboto-BoldItalic.ttf",
+            {"weight": BOLD, "style": ITALIC},
+        ),
+    ],
+)
+async def test_font_file_loaded(
+    widget: toga.Label,
+    widget_probe: Any,
+    font_family: str,
+    font_path: str,
+    font_kwargs,
+    capsys: pytest.CaptureFixture[str],
+    app_path: Path,
+):
+    """Custom fonts can be loaded and used."""
+    Font.register(
+        family=font_family,
+        path=app_path / font_path,
+        **font_kwargs,
+    )
+
+    # Update widget font family and other options if needed
+    widget.style.font_family = font_family
+    for prop, value in font_kwargs.items():
+        widget.style.update(
+            **{f"font_{kwarg}": value for kwarg, value in font_kwargs.items()}
+        )
+    await widget_probe.redraw(f"Using {font_family} {' '.join(font_kwargs.values())}")
+
+    # Check that font properties are updated
+    widget_probe.assert_font_family(font_family)
+    for prop, value in font_kwargs.items():
+        assert getattr(widget_probe.font, prop) == value
+
+    # Ensure the font was actually loaded.
+    assert "could not be found" not in capsys.readouterr().out
```

Comment:
Shouldn't this be "using system font as a fallback"?

Reply:
True - although we should probably check both just to make sure the test isn't failing due to a missing file.

Reply:
Ok - this turns out to be a lot more complex than I thought... notes below.

Reply:
I've now removed the mistaken stdout assertion.